### PR TITLE
engines/dfs: add DAOS File System (dfs) engine

### PR DESCRIPTION
**Pull request by johannlombardi**
DAOS is a new scale-out open-source object store.
See https://github.com/daos-stack/daos for more information.

This patch adds a new fio engine to support the filesystem layer built
on top of DAOS called DFS (DAOS File System). It supports asynchronous
read/write operations.

Signed-off-by: Johann Lombardi <johann.lombardi@intel.com>
Co-authored-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>
Co-authored-by: Vishwanath Venkatesan <vishwanath.venkatesan@intel.com>
Co-authored-by: Ioannis Galanis <ioannis.galanis@intel.com>

### CODE REVIEW

File: engines/dfs.c [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+	rc = dfs_mount(poh, coh, O_RDWR, &dfs);
+	if (rc) {
+		log_err("Failed to mount DFS namespace: %d\n", rc);
+		td_verror(td, EINVAL, "dfs_mount");
+		(void) daos_pool_disconnect(poh, NULL);
+		(void) daos_cont_close(coh, NULL);
+		return EINVAL;
+	}
+
+	/** Retrieve object class to use */
```
Comment (sitsofe) [inline engines/dfs.c:184]: Why two starts?

File: engines/dfs.c [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+	rc = daos_cont_close(coh, NULL);
+	if (rc)
+		log_err("failed to close container: %d\n", rc);
+	rc = daos_pool_disconnect(poh, NULL);
+	if (rc)
+		log_err("failed to disconnect pool: %d\n", rc);
+	rc = daos_fini();
+	if (rc)
+		log_err("failed to finalize daos: %d\n", rc);
+}
```
Comment (sitsofe) [inline engines/dfs.c:212]: Don't you want to return/set an error so callers know something went wrong?

**Review by sitsofe — CHANGES_REQUESTED** [review]
Some minor comments after a brief first pass review.

**Comment by sitsofe** [comment]
Oh and don't forget to add a prefix to your commit's title (see https://github.com/axboe/fio/commits/master/engines for ideas).

File: configure [hunk 3]
```diff
@@ -2179,6 +2182,32 @@ if test "$libnbd" != "no" ; then
 fi
 print_config "NBD engine" "$libnbd"
 
+##########################################
+# check for dfs (DAOS File System)
+if test "$dfs" != "yes" ; then
```
[the remark this answers is not in the document]
Comment (johannlombardi) [reply configure:2187]: fixed.

File: engines/dfs.c [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+		.group		= FIO_OPT_G_DFS,
+	},
+#endif
+	{
+		.name           = NULL,
+	},
+};
+
+static int
+daos_fio_global_init(struct thread_data *td)
```
[the remark this answers is not in the document]
Comment (johannlombardi) [reply engines/dfs.c:103]: fixed all occurrences.

File: engines/dfs.c (hunk 1 above)
Comment (johannlombardi) [reply engines/dfs.c:184]: Not sure I get this one. The code below just parses the object_class option (if any) and convert it from a string to our object class identifier enum. Could you please elaborate?

File: engines/dfs.c (hunk 2 above)
Comment (johannlombardi) [reply engines/dfs.c:212]: Good point. fixed.

**Comment by johannlombardi** [comment]
@sitsofe Thanks a lot for the review. I have addressed all your points except the one related to the two starts that I don't grok.
I have also updated the title of the PR to include a prefix. Let me know how it looks.

File: engines/dfs.c (hunk 1 above)
Comment (sitsofe) [reply engines/dfs.c:183]: Totally trivial: "/**" <- there's a second star :-)

File: engines/dfs.c [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+}
+
+static int daos_fio_init(struct thread_data *td)
+{
+	struct daos_data	*dd;
+	int			rc = 0;
+
+	pthread_mutex_lock(&daos_mutex);
+
+	/* Allocate space for DAOS-related data */
```
Comment (sitsofe) [inline engines/dfs.c:238]: I think you can skip this comment :-)

File: configure [hunk 6]
```diff
@@ -2179,6 +2183,33 @@ if test "$libnbd" != "no" ; then
...
+#include <daos.h>
+#include <daos_fs.h>
+
+int main(int argc, char **argv)
+{
+  daos_handle_t	poh;
+  daos_handle_t	coh;
+  dfs_t		*dfs;
+
+  (void)dfs_mount(poh, coh, O_RDWR, &dfs);
```
Comment (sitsofe) [inline configure:2200]: Fio generally a space after the bracket in this situation.

File: engines/dfs.c [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+	}
+
+	td->io_ops_data = dd;
+	num_threads++;
+out:
+	if (rc) {
+		if (dd && dd->io_us)
+			free(dd->io_us);
+		if (dd)
+			free(dd);
```
Comment (sitsofe) [inline engines/dfs.c:277]: Did you know that `free()` "works" on `NULL` as well (I've only recently come across https://stackoverflow.com/a/6084233 )?

File: engines/dfs.c [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+	if (td->io_ops_data == NULL)
+		return;
+
+	rc = daos_eq_destroy(dd->eqh, DAOS_EQ_DESTROY_FORCE);
+	if (rc < 0) {
+		log_err("failed to destroy event queue: %d\n", rc);
+		td_verror(td, rc, "daos_eq_destroy");
+	}
+
+	free(dd->io_us);
```
Comment (sitsofe) [inline engines/dfs.c:302]: Do you know `dd` must not be NULL?

**Review by sitsofe — COMMENTED** [review]
Looking good - only trivial comments from me. Can you squash your commits together so everything arrives as one?

File: engines/dfs.c (hunk 5 above)
Comment (johannlombardi) [reply engines/dfs.c:238]: done.

File: configure (hunk 6 above)
Comment (johannlombardi) [reply configure:2200]: done

File: engines/dfs.c (hunk 8 above)
Comment (johannlombardi) [reply engines/dfs.c:302]: Yes, it is tested at the beginning of this function.

File: engines/dfs.c (hunk 7 above)
Comment (johannlombardi) [reply engines/dfs.c:277]: I do. Let me remove the condition then if you think that it makes the code more readable.

File: engines/dfs.c (hunk 1 above)
Comment (johannlombardi) [reply engines/dfs.c]: ah, got it. Just different coding style in the DAOS project. Anyway, fixed.

**Comment by johannlombardi** [comment]
@sitsofe Patch updated as per your comments and all commits squashed together. Thanks.

**Comment by johannlombardi** [comment]
Just fixed a minor issue reported by a user. No further changes expected as far as I am concerned.

**Comment by johannlombardi** [comment]
@axboe Any chance to consider this PR for review/integration? Thanks in advance.

**Comment by johannlombardi** [comment]
I have added a fix for the compilation issue on msys2 and also tested the new dfs engine with different configurations on a sizable DAOS system. @axboe any chance to review/land this PR? Thanks in advance.

**Comment by axboe** [comment]
I'll take a look today.

**Comment by axboe** [comment]
Looks pretty clean to me. The only thing I noticed is that the daos_mutex is used a bit liberally for init, there should be no need to grab it before allocating and initializing a new data_data structure. It should only be needed once you get to checking for the init point, incrementing num_threads, etc. But that's minor.

**Comment by johannlombardi** [comment]
Thanks a lot @axboe. I will look into the mutex thing.

**Pull request merged**